src/psyclone/dynamo0p3.py

```diff
@@ -7514,6 +7514,90 @@ def gen_mark_halos_clean_dirty(self, parent):
+        :param dep_tools: an optional instance of DependencyTools so that the
+            caller can access any diagnostic messages detailing why the loop
+            iterations are not independent.
+        :type dep_tools: Optional[
+            :py:class:`psyclone.psyir.tools.DependencyTools]
+
+        :returns: True if the loop iterations are independent, False otherwise.
+        :rtype: bool
+
+        '''
+        from psyclone.psyir.tools import DependencyTools, DTCode
+        if not dep_tools:
+            dtools = DependencyTools()
+        else:
+            dtools = dep_tools
+
+        if self.loop_type in ["null", "colours"]:
+            # We know we can't parallelise these loops.
+            return False
+
+        try:
+            stat = dtools.can_loop_be_parallelised(
+                self, test_all_variables=test_all_variables,
+                signatures_to_ignore=signatures_to_ignore)
+            if stat:
+                return True
+        except (InternalError, KeyError):
+            # LFRic still has symbols that don't exist in the symbol_table
+            # until the gen_code() step, so the dependency analysis raises
+            # errors in some cases.
+            # TODO #1648 - "last_[halo]_cell_all_colours" for a coloured loop.
+            return True
+
+        # The generic DA says that this loop cannot be parallelised. However,
+        # that is often because it wrongly identifies field/operator array
+        # accesses (e.g. fld_proxy%data) as being scalars that are written to.
+        if self.loop_type in ["colour", "dof"]:
+            # This loop is either over cells of a single colour or DoFs. So
+            # long as the symbols that the DA is complaining about are fields
+            # or operators then this is safe to parallelise.
+            table = self.scope.symbol_table
+            all_fld_names = []
+            for type_desc in LFRicConstants().DATA_TYPE_MAP.values():
+                all_fld_names.append(type_desc["type"])
+
+            for msg in dtools.get_all_messages():
+                if msg.code != DTCode.WARN_SCALAR_WRITTEN_ONCE:
+                    return False
+                for name in msg.var_names:
+                    sym = table.lookup(name)
+                    # Ideally at this point we would compare sym.datatype with
+                    # LFRicTypes("RealFieldDataType") etc. but the LFRic PSy
+                    # layer doesn't use those types yet.
+                    if not (isinstance(sym.datatype, DataTypeSymbol) and
+                            sym.datatype.name in all_fld_names):
+                        # The Symbol is not a field or operator.
+                        return False
+                # It is safe to ignore this warning.
+                return True
+
+        elif self.loop_type == "":
+            # We can parallelise a non-coloured loop if it only updates
+            # quantities on discontinuous function spaces.
```

Comment:
A Kernel which updates a field on a continuous space must have a `GH_INC` argument (apart from the special case of kernels which redundantly write the same value to shared dofs). Therefore, checking `has_inc_arg` is all we need to do. I'll extend the comment to make this clear (because I'm bound to forget too).

Reply:
Tanks for the explanation